{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1FDX1UiaQXdZc502AzT0o9XQgSql4V_Ah","timestamp":1668388253055}],"collapsed_sections":["bXRLd25OKuCf","2zNo6g0JDEnu"]},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"bXRLd25OKuCf"},"source":["# **6.8701 | 6.8710 | HST.507**\n","\n","#**Fall 2022 Problem Set 5: Clustering Phylogenetic Trees**\n","\n","Due: Friday, November 19 at 11:59PM (submit copy of notebook to Stellar)"]},{"cell_type":"markdown","metadata":{"id":"TBFZarclPt4o"},"source":["#Load libraries\n"]},{"cell_type":"markdown","metadata":{"id":"qcntKfkrN-4T"},"source":["Feel free to import any additional libraries here."]},{"cell_type":"code","metadata":{"id":"IMjjMM_QPrUz","executionInfo":{"status":"ok","timestamp":1669865299874,"user_tz":300,"elapsed":3400,"user":{"displayName":"Hem Narayan Das Chaudhary","userId":"15689893306727550784"}},"colab":{"base_uri":"https://localhost:8080/"},"outputId":"64842179-3a68-4230-e1af-0f6bdb15e4a2"},"source":["from google.colab import files\n","!pip install ete3\n","from ete3 import Tree\n","\n","from sklearn.cluster import AgglomerativeClustering\n","import numpy as np\n","from itertools import combinations \n","\n"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Requirement already satisfied: ete3 in /usr/local/lib/python3.8/dist-packages (3.1.2)\n"]}]},{"cell_type":"markdown","metadata":{"id":"Ji7qMAIt4-qc"},"source":["# Clustering Phylogenetic Trees\n","\n","We've seen that phylogenetic tree algorithms can construct *many* different \"good\" putative evolutionary histories. An important and challenging problem is that of reducing a large number of trees to a smaller number of representative solutions. The objective of this assignment is to explore techniques for dealing\n","with large sets of different phylogenetic trees for the same data. This assignment is inspired by the paper [Statistically based postprocessing of phylogenetic analysis by clustering](https://www.ncbi.nlm.nih.gov/pubmed/12169558.) by Cara Stockham, Li-San Wang, and Tandy Warnow (Bioinformatics, Vol 18, Suppl. 1, 2002, pp. S285-S293). This problem is due\n","to Ran Libeskind-Hadas. The full details are available at the URL below:\n","\n","http://www.cs.hmc.edu/~hadas/mitcompbio/treedistance.html"]},{"cell_type":"markdown","metadata":{"id":"MIzMHwWgLRDC"},"source":["You will need to download the following [files](https://www.dropbox.com/sh/94bcsf8hxer11fq/AABqiZjDqHgMHbwJhor7hR5Ja?dl=0) and upload them to colab using the following code block."]},{"cell_type":"code","metadata":{"id":"jRrT6-qvLW1d","colab":{"base_uri":"https://localhost:8080/","height":73},"executionInfo":{"status":"ok","timestamp":1669859317021,"user_tz":300,"elapsed":6779,"user":{"displayName":"Hem Narayan Das Chaudhary","userId":"15689893306727550784"}},"outputId":"45c0074d-9ba6-4c1e-ecd4-15f9ba2b9425"},"source":["caesal = files.upload()\n","from Caesal import *"],"execution_count":null,"outputs":[{"output_type":"display_data","data":{"text/plain":["<IPython.core.display.HTML object>"],"text/html":["\n","     <input type=\"file\" id=\"files-78fea81e-6fa5-49f8-ba0a-776ba9a8a740\" name=\"files[]\" multiple disabled\n","        style=\"border:none\" />\n","     <output id=\"result-78fea81e-6fa5-49f8-ba0a-776ba9a8a740\">\n","      Upload widget is only available when the cell has been executed in the\n","      current browser session. Please rerun this cell to enable.\n","      </output>\n","      <script>// Copyright 2017 Google LLC\n","//\n","// Licensed under the Apache License, Version 2.0 (the \"License\");\n","// you may not use this file except in compliance with the License.\n","// You may obtain a copy of the License at\n","//\n","//      http://www.apache.org/licenses/LICENSE-2.0\n","//\n","// Unless required by applicable law or agreed to in writing, software\n","// distributed under the License is distributed on an \"AS IS\" BASIS,\n","// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n","// See the License for the specific language governing permissions and\n","// limitations under the License.\n","\n","/**\n"," * @fileoverview Helpers for google.colab Python module.\n"," */\n","(function(scope) {\n","function span(text, styleAttributes = {}) {\n","  const element = document.createElement('span');\n","  element.textContent = text;\n","  for (const key of Object.keys(styleAttributes)) {\n","    element.style[key] = styleAttributes[key];\n","  }\n","  return element;\n","}\n","\n","// Max number of bytes which will be uploaded at a time.\n","const MAX_PAYLOAD_SIZE = 100 * 1024;\n","\n","function _uploadFiles(inputId, outputId) {\n","  const steps = uploadFilesStep(inputId, outputId);\n","  const outputElement = document.getElementById(outputId);\n","  // Cache steps on the outputElement to make it available for the next call\n","  // to uploadFilesContinue from Python.\n","  outputElement.steps = steps;\n","\n","  return _uploadFilesContinue(outputId);\n","}\n","\n","// This is roughly an async generator (not supported in the browser yet),\n","// where there are multiple asynchronous steps and the Python side is going\n","// to poll for completion of each step.\n","// This uses a Promise to block the python side on completion of each step,\n","// then passes the result of the previous step as the input to the next step.\n","function _uploadFilesContinue(outputId) {\n","  const outputElement = document.getElementById(outputId);\n","  const steps = outputElement.steps;\n","\n","  const next = steps.next(outputElement.lastPromiseValue);\n","  return Promise.resolve(next.value.promise).then((value) => {\n","    // Cache the last promise value to make it available to the next\n","    // step of the generator.\n","    outputElement.lastPromiseValue = value;\n","    return next.value.response;\n","  });\n","}\n","\n","/**\n"," * Generator function which is called between each async step of the upload\n"," * process.\n"," * @param {string} inputId Element ID of the input file picker element.\n"," * @param {string} outputId Element ID of the output display.\n"," * @return {!Iterable<!Object>} Iterable of next steps.\n"," */\n","function* uploadFilesStep(inputId, outputId) {\n","  const inputElement = document.getElementById(inputId);\n","  inputElement.disabled = false;\n","\n","  const outputElement = document.getElementById(outputId);\n","  outputElement.innerHTML = '';\n","\n","  const pickedPromise = new Promise((resolve) => {\n","    inputElement.addEventListener('change', (e) => {\n","      resolve(e.target.files);\n","    });\n","  });\n","\n","  const cancel = document.createElement('button');\n","  inputElement.parentElement.appendChild(cancel);\n","  cancel.textContent = 'Cancel upload';\n","  const cancelPromise = new Promise((resolve) => {\n","    cancel.onclick = () => {\n","      resolve(null);\n","    };\n","  });\n","\n","  // Wait for the user to pick the files.\n","  const files = yield {\n","    promise: Promise.race([pickedPromise, cancelPromise]),\n","    response: {\n","      action: 'starting',\n","    }\n","  };\n","\n","  cancel.remove();\n","\n","  // Disable the input element since further picks are not allowed.\n","  inputElement.disabled = true;\n","\n","  if (!files) {\n","    return {\n","      response: {\n","        action: 'complete',\n","      }\n","    };\n","  }\n","\n","  for (const file of files) {\n","    const li = document.createElement('li');\n","    li.append(span(file.name, {fontWeight: 'bold'}));\n","    li.append(span(\n","        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n","        `last modified: ${\n","            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n","                                    'n/a'} - `));\n","    const percent = span('0% done');\n","    li.appendChild(percent);\n","\n","    outputElement.appendChild(li);\n","\n","    const fileDataPromise = new Promise((resolve) => {\n","      const reader = new FileReader();\n","      reader.onload = (e) => {\n","        resolve(e.target.result);\n","      };\n","      reader.readAsArrayBuffer(file);\n","    });\n","    // Wait for the data to be ready.\n","    let fileData = yield {\n","      promise: fileDataPromise,\n","      response: {\n","        action: 'continue',\n","      }\n","    };\n","\n","    // Use a chunked sending to avoid message size limits. See b/62115660.\n","    let position = 0;\n","    do {\n","      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n","      const chunk = new Uint8Array(fileData, position, length);\n","      position += length;\n","\n","      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n","      yield {\n","        response: {\n","          action: 'append',\n","          file: file.name,\n","          data: base64,\n","        },\n","      };\n","\n","      let percentDone = fileData.byteLength === 0 ?\n","          100 :\n","          Math.round((position / fileData.byteLength) * 100);\n","      percent.textContent = `${percentDone}% done`;\n","\n","    } while (position < fileData.byteLength);\n","  }\n","\n","  // All done.\n","  yield {\n","    response: {\n","      action: 'complete',\n","    }\n","  };\n","}\n","\n","scope.google = scope.google || {};\n","scope.google.colab = scope.google.colab || {};\n","scope.google.colab._files = {\n","  _uploadFiles,\n","  _uploadFilesContinue,\n","};\n","})(self);\n","</script> "]},"metadata":{}},{"output_type":"stream","name":"stdout","text":["Saving Caesal.py to Caesal.py\n"]}]},{"cell_type":"markdown","metadata":{"id":"7asIx9MH6aAb"},"source":["A) Implement the *Robinson-Foulds* distance metric."]},{"cell_type":"code","metadata":{"id":"T0ca4hFfLHop"},"source":["\"\"\"\n","Returns distance between two phylogenetic trees\n","\n","Arguments:\n","    tree1 - first list of lists or tuple of tuples, with integer member elements\n","    tree2 - second list of lists or tuple of tuples, with integer member elements\n","Returns:\n","    dist - integer distance between 'tree1' and 'tree2'\n","    \"\"\"\n","def distance(tree1, tree2):\n","    ### YOUR CODE HERE ###\n","    t1=Tree(str(tree1)+';')\n","    t2=Tree(str(tree2)+';')\n","    dist=int(t1.robinson_foulds(t2)[0]/2)\n","    return dist"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"6h-WJvizNroV"},"source":["These are the test cases for your `distance` function."]},{"cell_type":"code","metadata":{"id":"OFq-9KVpNH3s"},"source":["def test_distance():\n","    T1 = (0, (1, (2, (3, 4))))\n","    T2 = (0, (1, (3, (2, 4))))\n","    print(\"Expected: 1\\tOutput: \" + str(distance(T1, T2)))\n","    T3 = (0, ((1, (2, 3)), (7, (6, (4, 5)))))\n","    T4 = (0, ((2, (1, 3)), (6, (4, (5, 7)))))\n","    print(\"Expected: 3\\tOutput: \" + str(distance(T3, T4)))\n","    print(\"Expected: 8\\tOutput: \" + str(distance(treeList[0], treeList[42])))\n","    print(\"Expected: 3\\tOutput: \" + str(distance(treeList[57], treeList[77])))\n","    print(\"Expected: 1\\tOutput: \" + str(distance(treeList[192], treeList[200])))\n","    print(\"Expected: 4\\tOutput: \" + str(distance(treeList[386], treeList[234])))"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"0m1y4XWwNV1H","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1669867198599,"user_tz":300,"elapsed":10,"user":{"displayName":"Hem Narayan Das Chaudhary","userId":"15689893306727550784"}},"outputId":"61fef43e-cc29-4561-cf6e-481d61f1a4ac"},"source":["test_distance()"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Expected: 1\tOutput: 1\n","Expected: 3\tOutput: 3\n","Expected: 8\tOutput: 8\n","Expected: 3\tOutput: 3\n","Expected: 1\tOutput: 1\n","Expected: 4\tOutput: 4\n"]}]},{"cell_type":"markdown","metadata":{"id":"1eO6gwjMOMch"},"source":["B) You will use a clustering algorithm to partition distinct phylogenetic trees into clusters of trees where the trees in a given cluster are similar with respect to *Robinson-Foulds* distance. (Hint: take note of their hint for speeding up the *Robinson-Foulds* distance calculations!)"]},{"cell_type":"code","metadata":{"id":"OX3xFvPKOXDP"},"source":["\"\"\"\n","Return each clustering step from 'maxClusters' to 'minClusters' size from input 'treeList'\n","\n","Arguments:\n","    treeList - list of all trees to cluster\n","    minClusters - integer size of minimum number of clusters\n","    maxClusters - integer size of maximum number of clusters\n","Returns:\n","    clusterings - list of clusters with information on numbers of trees, diameters (Robinson-Foulds distance), and average diameter\n","        [[[numbers of trees], [diameters], average diameter],   <- max number of clusters\n","         .\n","         .\n","         .\n","         [[numbers of trees], [diameters], average diameter]]   <- min number of clusters\n","\"\"\"\n","def cluster(treeList, minClusters, maxClusters):\n","    ### YOUR CODE HERE ###\n","    #generate 2D distance\n","    n=len(treeList)\n","    N=np.zeros((n,n))\n","    for k in range(1,n):#calculating distance matrix\n","      for j in range(k,n):\n","        i=j-k\n","        N[i,j]=distance(treeList[i],treeList[j])\n","        N[j,i]=N[i,j]\n","    X = np.array(N)\n","    clusterings=[]\n","    for l in range(maxClusters,minClusters,-1): #for varying number of clusters\n","          clust = AgglomerativeClustering(n_clusters=l,linkage='average').fit(X) #cluster distance\n","          clustl=clust.labels_\n","          numtrees=[]\n","          diameter=[]\n","          avgdiam=0\n","          for m in range(l): #for every cluster\n","            clustindex=[]\n","            for o in range(len(clustl)): #keep track of index of tree for its cluster\n","              if clustl[o]==m:\n","                clustindex.append(o)\n","            numtrees.append(len(clustindex)) #number of trees in each clusters\n","            dist_for_eachC=[]\n","            pwc=list(combinations(clustindex,2)) #pairwise combinations\n","            for p in pwc:\n","              dist_for_eachC.append(N[p]) #retriving distance for pariwise combination\n","            diameter.append(max(dist_for_eachC)) #max among distance \n","            avgdiam=np.average(diameter) #average of the max distances \n","          clusterings.append([numtrees,diameter,avgdiam]) #putting things together\n","\n","    return clusterings"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"BlpZsYxJOg1R"},"source":["These are the test cases for your `cluster` function."]},{"cell_type":"code","metadata":{"id":"9nVeftQJOkeG"},"source":["def test_cluster():\n","    clusters = [[[3, 6, 3, 6, 2], [1, 2, 1, 2, 1], 1.4],\n","                [[9, 3, 6, 2], [2, 1, 2, 1], 1.5],\n","                [[9, 9, 2], [2, 3, 1], 2],\n","                [[18, 2], [5, 1], 3],\n","                [[20], [7], 7]]\n","    for i, clust in enumerate(cluster(treeList[:20], 1, 5)):\n","        print(\"Expected: \" + str(clusters[i]) + \"\\nOutput:   \" + str(clust))"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"6R3lDTC8JfEW","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1669867215659,"user_tz":300,"elapsed":784,"user":{"displayName":"Hem Narayan Das Chaudhary","userId":"15689893306727550784"}},"outputId":"996af7d4-b102-44cc-ac71-9e2d49f86dd2"},"source":["test_cluster()"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Expected: [[3, 6, 3, 6, 2], [1, 2, 1, 2, 1], 1.4]\n","Output:   [[6, 3, 6, 2, 3], [2.0, 1.0, 2.0, 1.0, 1.0], 1.4]\n","Expected: [[9, 3, 6, 2], [2, 1, 2, 1], 1.5]\n","Output:   [[9, 3, 6, 2], [2.0, 1.0, 2.0, 1.0], 1.5]\n","Expected: [[9, 9, 2], [2, 3, 1], 2]\n","Output:   [[9, 2, 9], [3.0, 1.0, 2.0], 2.0]\n","Expected: [[18, 2], [5, 1], 3]\n","Output:   [[18, 2], [5.0, 1.0], 3.0]\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.8/dist-packages/sklearn/cluster/_agglomerative.py:542: ClusterWarning: scipy.cluster: The symmetric non-negative hollow observation matrix looks suspiciously like an uncondensed distance matrix\n","  out = hierarchy.linkage(X, method=linkage, metric=affinity)\n","/usr/local/lib/python3.8/dist-packages/sklearn/cluster/_agglomerative.py:542: ClusterWarning: scipy.cluster: The symmetric non-negative hollow observation matrix looks suspiciously like an uncondensed distance matrix\n","  out = hierarchy.linkage(X, method=linkage, metric=affinity)\n","/usr/local/lib/python3.8/dist-packages/sklearn/cluster/_agglomerative.py:542: ClusterWarning: scipy.cluster: The symmetric non-negative hollow observation matrix looks suspiciously like an uncondensed distance matrix\n","  out = hierarchy.linkage(X, method=linkage, metric=affinity)\n","/usr/local/lib/python3.8/dist-packages/sklearn/cluster/_agglomerative.py:542: ClusterWarning: scipy.cluster: The symmetric non-negative hollow observation matrix looks suspiciously like an uncondensed distance matrix\n","  out = hierarchy.linkage(X, method=linkage, metric=affinity)\n"]}]},{"cell_type":"markdown","metadata":{"id":"S0hvjn5JMGLu"},"source":["C) You will implement a general consensus algorithm that will allow you to find a consensus tree for each cluster and measure the quality of the cluster by its \"specificity\" (how close it is to being a binary tree)."]},{"cell_type":"code","metadata":{"id":"QfZWz-sTMRgP"},"source":["\"\"\"\n","Calculate the specificity of 'cluster' given a target 'threshold'\n","\n","Arguments:\n","    cluster - list of trees\n","    threshold - float value of threshold\n","Returns:\n","    spec - float specificity of 'cluster' given 'threshold'\n","\"\"\"\n","def specificity(cluster, threshold):\n","    ### YOUR CODE HERE ###\n","    return spec"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"CrgV0M7hMe9y"},"source":["These are the test cases for your `specificity` function."]},{"cell_type":"code","metadata":{"id":"pqCbd5jjMmFr"},"source":["def test_specificity():\n","    print(\"Expected: 0.77083\\tOutput: \" + str(specificity(treeList, 1.0)))\n","    print(\"Expected: 0.91667\\tOutput: \" + str(specificity(treeList, 0.51)))"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"pJsDVVfHMmIA"},"source":["test_specificity()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2zNo6g0JDEnu"},"source":["# (6.878 only) Coalescent simulation\n","\n","In this problem, we will simulate the coalescent process. Recall that this is the time-reversal of the Wright-Fisher process."]},{"cell_type":"markdown","metadata":{"id":"lMSXGwj1vD02"},"source":["A) Write a program to simulate the coalescent process on a population of $2N$ alleles. Track the times of coalescent events starting from the initial generation until all alleles coalesce to a single ancestor. If we are tracking $k$ lineages, you should report $k - 1$ coalescent events.\n","\n","Recall that the Wright-Fisher process assumes each allele in the next generation is sampled independently from all alleles in the current generation. We are now interested in the reverse, so we instead need to sample parents in the previous generation uniformly at random with replacement. Note that we are interested in the identities of the parents and not their ancestral alleles.\n","\n","Run 1,000 trials with a population size of $N = 500$. Report the mean and standard deviation of the number of generations between coalescent events for $k = 2$, $3$, $4$ and $5$ lineages."]},{"cell_type":"code","metadata":{"id":"TiSyRzAEvchQ"},"source":["### YOUR CODE HERE ###"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"GbHCUIC3vf6r"},"source":["**Answer here:**\n","\n","|Lineages|Mean|Standard Deviation|\n","|-|-|-|\n","|2|||\n","|3|||\n","|4|||\n","|5|||"]},{"cell_type":"markdown","metadata":{"id":"DGTcFFGmwGDs"},"source":["B) Recall the waiting time between coalescent events is approximately exponentially distributed with parameter $\\lambda$. For each value of $k$, what is the value of $\\lambda$ given $N = 500$?\n","\n","Assuming this distribution, the mean waiting time and its standard deviation are both $\\frac{1}{\\lambda}$. How do these expected values compare to your observed values? If your observed values are different, give an explanation of what could have caused the differences."]},{"cell_type":"markdown","metadata":{"id":"JL5NM9dGwicH"},"source":["**Answer here:**\n","\n","|Lineages|$\\lambda$|$\\frac{1}{\\lambda}$|\n","|-|-|-|\n","|2|||\n","|3|||\n","|4|||\n","|5|||"]},{"cell_type":"markdown","metadata":{"id":"sf-esvM5wwDl"},"source":["C) Extend your simulator to model sexual reproduction.\n","\n","Assume a fixed number of females $F$ (and therefore $M = N - F$ males) in each generation and that each chromosome in the next generation is selected in the following way: sample a male and female to mate uniformly at random, then sample one of the two alleles uniformly at random. Your simulation should do\n","the reverse: sample a father and mother and then pick one at random as the ancestor for each allele.\n","\n","Run 1,000 trials with $F = 100$ and $M = 400$. Do your results agree with the coalescent approximation? Justify your results.\n","\n","Can you extend the coalescent approximation to more accurately reflect this model of sexual reproduction? Do your results agree with this new approximation?"]},{"cell_type":"code","metadata":{"id":"jTNLD7PmxPz4"},"source":["### YOUR CODE HERE ###"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ktNa0jnvxRFJ"},"source":["**Answer here:**\n","\n","|Lineages|Mean|Standard Deviation|\n","|-|-|-|\n","|2|||\n","|3|||\n","|4|||\n","|5|||"]}]}